lib/repl.js

```diff
@@ -258,21 +264,84 @@ function REPLServer(prompt,
       sawSIGINT = false;
     }
 
+    self._currentStringLiteral = null;
     self.bufferedCommand = '';
     self.lines.level = [];
     self.displayPrompt();
   });
 
+  function parseForStringLiteral(line) {
+    var previous = null, current = null;
+
+    for (var i = 0; i < line.length; i += 1) {
+      if (previous === '\\') {
+        // if it is a valid escaping, then skip processing
+        previous = current;
+        continue;
+      }
+
+      current = line.charAt(i);
+      if (current === self._currentStringLiteral) {
+        self._currentStringLiteral = null;
+      } else if (current === '\'' ||
+                 current === '"' &&
+                 self._currentStringLiteral === null) {
+        self._currentStringLiteral = current;
+      }
+      previous = current;
+    }
+  }
+
+  function getFinisherFunction(cmd, defaultFn) {
+    if ((self._currentStringLiteral === null &&
+         cmd.charAt(cmd.length - 1) === '\\') ||
+        (self._currentStringLiteral !== null &&
+         cmd.charAt(cmd.length - 1) !== '\\')) {
+
+      // If the line continuation is used outside string literal or if the
+      // string continuation happens with out line continuation, then fail hard
+      return function(e, ret) {
+        var error = e instanceof Recoverable ? e.err : e;
+        var args = [error];
+
+        if (arguments.length === 2) {
+          // adding second argument only if it is actually passed. Otherwise
+          // `undefined` will be printed when invalid REPL commands are used.
+          args.push(ret);
+        }
+
+        return defaultFn.apply(this, args);
+      };
+    }
+    return defaultFn;
+  }
+
   self.on('line', function(cmd) {
     debug('line %j', cmd);
     sawSIGINT = false;
     var skipCatchall = false;
+    var finisherFn = finish;
 
     // leading whitespaces in template literals should not be trimmed.
     if (self._inTemplateLiteral) {
       self._inTemplateLiteral = false;
     } else {
-      cmd = trimWhitespace(cmd);
+      var wasWithinStrLiteral = self._currentStringLiteral !== null;
+      parseForStringLiteral(cmd);
+      var isWithinStrLiteral = self._currentStringLiteral !== null;
+
+      if (!wasWithinStrLiteral && !isWithinStrLiteral) {
+        // Current line has nothing to do with String literals, trim both ends
+        cmd = cmd.replace(/^\s+|\s+$/g, '');
+      } else if (wasWithinStrLiteral && !isWithinStrLiteral) {
+        // was part of a string literal, but it is over now, trim only the end
+        cmd = cmd.replace(/\s+$/, '');
+      } else if (isWithinStrLiteral && !wasWithinStrLiteral) {
+        // was not part of a string literal, but it is now, trim only the start
+        cmd = cmd.replace(/^\s+/, '');
+      }
```

Comment:
Just yesterday I learned about `String.prototype.trimLeft()` and `String.prototype.trimRight()`. These are nonstandard, but there's some discussion about standardizing on the es-discuss mailing list, so I think we're safe to use them. Also, on the first case, let's use standard `String.prototype.trim()` :)